File: src/services/offer.ts

import { supabase } from "@/integrations/supabase/client";
import { handleFetchError } from "./common";

interface OfferFilters {
  status?: string;
  supplier_id?: string;
  [key: string]: any;
}

export const offerService = {
  // Récupérer toutes les offres
  getAllOffers: async (filters: OfferFilters = {}) => {
    try {
      let query = supabase
        .from('offers')
        .select(`
          *,
          profiles:supplier_id (
            business_name,
            phone,
            email
          )
        `)
        .order('created_at', { ascending: false });
      
      // Apply filters if provided
      if (filters.status) {
        query = query.eq('status', filters.status);
      }
      if (filters.supplier_id) {
        query = query.eq('supplier_id', filters.supplier_id);
      }
      
      const { data, error } = await query;
      
      if (error) {
        console.error("Error fetching offers:", error);
        return null;
      }
      
      return { offers: data || [] };
    } catch (error) {
      return handleFetchError(error);
    }
  },
  
  // Récupérer une offre par ID
  getOfferById: async (id: string) => {
    try {
      const { data, error } = await supabase
        .from('offers')
        .select(`
          *,
          profiles:supplier_id (
            business_name,
            phone,
            email
          )
        `)
        .eq('id', id)
        .single();
      
      if (error) {
        console.error("Error fetching offer:", error);
        return null;
      }
      
      return { offer: data };
    } catch (error) {
      return handleFetchError(error);
    }
  },
  
  // Créer une nouvelle offre
  createOffer: async (offerData: any) => {
    try {
      const { data: { user } } = await supabase.auth.getUser();
      
      if (!user) {
        return { error: "Utilisateur non authentifié" };
      }

      const { data, error } = await supabase
        .from('offers')
        .insert({
          ...offerData,
          supplier_id: user.id,
        })
        .select()
        .single();
      
      if (error) {
        console.error("Error creating offer:", error);
        return { error: error.message };
      }
      
      return { 
        offer: data, 
        message: "Offre créée avec succès" 
      };
    } catch (error) {
      return handleFetchError(error);
    }
  },
  
  // Mettre à jour une offre
  updateOffer: async (id: string, offerData: any) => {
    try {
      const { data, error } = await supabase
        .from('offers')
        .update(offerData)
        .eq('id', id)
        .select()
        .single();
      
      if (error) {
        console.error("Error updating offer:", error);
        return { error: error.message };
      }
      
      return { 
        offer: data, 
        message: "Offre mise à jour avec succès" 
      };
    } catch (error) {
      return handleFetchError(error);
    }
  },
  
  // Supprimer une offre
  deleteOffer: async (id: string) => {
    try {
      const { error } = await supabase
        .from('offers')
        .delete()
        .eq('id', id);
      
      if (error) {
        console.error("Error deleting offer:", error);
        return { error: error.message };
      }
      
      return { message: "Offre supprimée avec succès" };
    } catch (error) {
      return handleFetchError(error);
    }
  },
};

File: backend/handlers/offer.go

package handlers

import (
	"context"
	"net/http"
	"time"

	"elsaidaliya/middleware"
	"elsaidaliya/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

var offerCollection *mongo.Collection

// InitOfferHandlers initializes handlers for offer management
func InitOfferHandlers(db *mongo.Database) {
	offerCollection = db.Collection("offers")
}

// GetAllOffers fetches all offers
func GetAllOffers(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Define options to sort by createdAt DESC
	findOptions := options.Find()
	findOptions.SetSort(bson.M{"createdAt": -1})

	// Get filter from query params
	filter := bson.M{}
	if supplierID := c.Query("supplier"); supplierID != "" {
		objectID, err := primitive.ObjectIDFromHex(supplierID)
		if err == nil {
			filter["supplierID"] = objectID
		}
	}

	cursor, err := offerCollection.Find(ctx, filter, findOptions)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors de la récupération des offres"})
		return
	}
	defer cursor.Close(ctx)

	var offers []models.Offer
	if err = cursor.All(ctx, &offers); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors du traitement des offres"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"offers": offers})
}

// GetOfferByID fetches a single offer by ID
func GetOfferByID(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	offerID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "ID d'offre invalide"})
		return
	}

	var offer models.Offer
	err = offerCollection.FindOne(ctx, bson.M{"_id": offerID}).Decode(&offer)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Offre non trouvée"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors de la récupération de l'offre"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"offer": offer})
}

// CreateOffer creates a new offer
func CreateOffer(c *gin.Context) {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Utilisateur non authentifié"})
		return
	}

	var newOffer models.Offer
	if err := c.ShouldBindJSON(&newOffer); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Données invalides", "details": err.Error()})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Set the supplier ID to the current user
	newOffer.ID = primitive.NewObjectID()
	newOffer.SupplierID = userID
	newOffer.CreatedAt = time.Now()
	newOffer.UpdatedAt = time.Now()

	// Insert offer
	_, err = offerCollection.InsertOne(ctx, newOffer)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors de la création de l'offre"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"message": "Offre créée avec succès",
		"offer": newOffer,
	})
}

// UpdateOffer updates an existing offer
func UpdateOffer(c *gin.Context) {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Utilisateur non authentifié"})
		return
	}

	offerID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "ID d'offre invalide"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Check if offer exists and belongs to user
	var existingOffer models.Offer
	err = offerCollection.FindOne(ctx, bson.M{
		"_id": offerID,
		"supplierID": userID,
	}).Decode(&existingOffer)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Offre non trouvée ou vous n'avez pas l'autorisation de la modifier"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors de la vérification de l'offre"})
		return
	}

	// Bind the update data
	var updateData models.Offer
	if err := c.ShouldBindJSON(&updateData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Données invalides"})
		return
	}

	// Create update document
	update := bson.M{
		"$set": bson.M{
			"title":       updateData.Title,
			"description": updateData.Description,
			"price":       updateData.Price,
			"imageUrl":    updateData.ImageURL,
			"expiresAt":   updateData.ExpiresAt,
			"updatedAt":   time.Now(),
		},
	}

	// Update the offer
	_, err = offerCollection.UpdateOne(ctx, bson.M{"_id": offerID}, update)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors de la mise à jour de l'offre"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Offre mise à jour avec succès"})
}

// DeleteOffer deletes an offer
func DeleteOffer(c *gin.Context) {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Utilisateur non authentifié"})
		return
	}

	offerID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "ID d'offre invalide"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Only allow deletion of user's own offers (or admin can delete any)
	userRole, _ := c.Get("userRole")
	filter := bson.M{"_id": offerID}
	if userRole != "admin" {
		filter["supplierID"] = userID
	}

	result, err := offerCollection.DeleteOne(ctx, filter)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erreur lors de la suppression de l'offre"})
		return
	}

	if result.DeletedCount == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Offre non trouvée ou vous n'avez pas l'autorisation de la supprimer"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Offre supprimée avec succès"})
}
